src/lenskit/reranking/fair.py

```diff
@@ -0,0 +1,223 @@
+        m_list_ = np.clip(m_list_, 0, k_vals).astype(int)
+        return m_list_
+
+    def _compute_blocks(self, m_list_):
+        max_m = int(m_list_[-1])
+        if max_m == 0:
+            return []
+
+        # find the positions where m increases to form blocks
+        # (last value of each block increases by 1)
+        change_points = np.flatnonzero(np.diff(m_list_, prepend=0)) + 1
+        block_sizes = np.diff(change_points, prepend=0)
+        return block_sizes
+
+    def _compute_rejection_prob(self, k, p, alpha_c):
+        """
+        Algorithm 1 (AdjustSignificance), exactly:
+        - compute m(pos) from Binomial PPF at alpha_c
+        - compute blocks b(j)
+        - for j=1..m(k):
+            S = conv(S, Bin(b(j), p))
+            S[j-1] = 0     # prune bucket j-1
+        Returns rejection probability = 1- sum(S).
+        """
+        m_list_ = self._compute_m_list(k, p, alpha_c)
+        m_blocks = self._compute_blocks(m_list_)
+        S = np.array([1.0], dtype=np.float64)
+
+        if len(m_blocks) > 0:
+            for j, block_size in enumerate(m_blocks, start=1):
+                if block_size not in self.pmf_cache:
+                    self.pmf_cache[block_size] = binom.pmf(np.arange(block_size + 1), block_size, p)
+
+                S = np.convolve(self.pmf_cache[block_size], S)
+                # set the next index (j-1) to zero since the last element of each block
+                # increases by 1 and now we need at least j protected up to this point
+                S[j - 1] = 0
+
+        return float(1 - S.sum())
+
+    def binary_search_significance(self, k, p, alpha, tolerance=1e-10, max_iter=100):
+        low, high = 0, 1
+
+        for _ in range(max_iter):
+            alpha_c = (low + high) / 2
+            rejection_prob = self._compute_rejection_prob(k, p, alpha_c)
+
+            if rejection_prob > alpha:
+                high = alpha_c
+            else:
+                low = alpha_c
+            if abs(rejection_prob - alpha) < tolerance or high - low < tolerance:
+                break
+
+        alpha_c = (low + high) / 2
+        return alpha_c
+
+    @override
+    def train(self, data: Dataset, options: TrainingOptions = TrainingOptions()):
+        """
+        Precompute adjusted alpha and m-table for the configured (k,p,alpha)
+
+        Training computes the (multiple-tests–corrected) prefix requirements m[i]
+        so that every prefix of size i has at least m[i] protected items with
+        significance α (after correction).
+
+        Args:
+            protected attributes:
+                The item entity table must have a scalar attribute ``protected`` that is ``True``
+                if the item is in the protected group and ``False`` otherwise.
+        """
+        self.pmf_cache = {}
+
+        log = _log.bind(alpha=self.config.alpha, k_max=self.config.k)
+        timer = Stopwatch()
+        log.info("begining FA*IR Tuning")
```

Comment:
we aren't really tuning anything, are we? Maybe “computing FA*IR thresholds” or something like that?